src/models/edge.ts

```diff
@@ -400,15 +402,56 @@ abstract class Edge<N extends INodeBase, E extends IEdgeBase> extends Subject im
   }
 
   setState(state: number): void;
+  setState(state: IGraphObjectStateParameters): void;
   setState(callback: (edge: IEdge<N, E>) => number): void;
-  setState(arg: number | ((edge: IEdge<N, E>) => number)): void {
+  setState(callback: (edge: IEdge<N, E>) => IGraphObjectStateParameters): void;
+  setState(
+    arg:
+      | number
+      | IGraphObjectStateParameters
+      | ((edge: IEdge<N, E>) => number)
+      | ((edge: IEdge<N, E>) => IGraphObjectStateParameters),
+  ): void {
+    let result: number | IGraphObjectStateParameters;
+
     if (isFunction(arg)) {
-      this._state = (arg as (edge: IEdge<N, E>) => number)(this);
+      result = (arg as (edge: IEdge<N, E>) => number | IGraphObjectStateParameters)(this);
     } else {
-      this._state = arg as number;
+      result = arg;
     }
+
+    if (typeof result === 'number') {
+      this._state = result;
+    } else if (typeof result === 'object') {
+      const options = result.options;
+
+      if (options && options.isToggle) {
+        this._toggleState(result.state);
+      } else {
+        this._state = result.state;
+      }
```

Comment:
Either have `this._state = <function that handles options in general>(results.state)` where function doesn't mutate the state, but returns what the new state should be, or keep this as is, but still change the `toggleState` to not mutate the `this._state` because I don't see a reason why it should. The best functions are the clean ones, which do not mutate state (if it is possible to make them).